receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go

```diff
@@ -16,54 +16,270 @@ package k8sapiserver
-		return nil
+	for _, opt := range options {
+		opt(k)
+	}
+
+	if k.k8sClient == nil {
+		return nil, errors.New("failed to start k8sapiserver because k8sclient is nil")
 	}
 
-	return k
+	if err := k.init(); err != nil {
+		return nil, fmt.Errorf("fail to initialize k8sapiserver, err: %v", err)
+	}
+
+	return k, nil
 }
 
 // GetMetrics returns an array of metrics
 func (k *K8sAPIServer) GetMetrics() []pdata.Metrics {
-	// TODO: add the logic to generate the metrics
 	var result []pdata.Metrics
+
+	//don't emit metrics if the cluster name is not detected
+	clusterName := k.clusterNameProvider.GetClusterName()
+	if clusterName == "" {
+		k.logger.Warn("Failed to detect cluster name. Drop all metrics")
+		return result
+	}
+
+	if k.leading {
+		k.logger.Info("collect data from K8s API Server...")
+		timestampNs := strconv.FormatInt(time.Now().UnixNano(), 10)
+		client := k.k8sClient
+
+		fields := map[string]interface{}{
+			"cluster_failed_node_count": client.Node.ClusterFailedNodeCount(),
+			"cluster_node_count":        client.Node.ClusterNodeCount(),
+		}
+		attributes := map[string]string{
+			ci.ClusterNameKey: clusterName,
+			ci.MetricType:     ci.TypeCluster,
+			ci.Timestamp:      timestampNs,
+			ci.Version:        "0",
+		}
+		if k.nodeName != "" {
+			attributes["NodeName"] = k.nodeName
+		}
+		md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+		result = append(result, md)
+
+		for service, podNum := range client.Ep.ServiceToPodNum() {
+			fields := map[string]interface{}{
+				"service_number_of_running_pods": podNum,
+			}
+			attributes := map[string]string{
+				ci.ClusterNameKey: clusterName,
+				ci.MetricType:     ci.TypeClusterService,
+				ci.Timestamp:      timestampNs,
+				ci.TypeService:    service.ServiceName,
+				ci.K8sNamespace:   service.Namespace,
+				ci.Version:        "0",
+			}
+			if k.nodeName != "" {
+				attributes["NodeName"] = k.nodeName
+			}
+			md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+			result = append(result, md)
+		}
+
+		for namespace, podNum := range client.Pod.NamespaceToRunningPodNum() {
+			fields := map[string]interface{}{
+				"namespace_number_of_running_pods": podNum,
+			}
+			attributes := map[string]string{
+				ci.ClusterNameKey: clusterName,
+				ci.MetricType:     ci.TypeClusterNamespace,
+				ci.Timestamp:      timestampNs,
+				ci.K8sNamespace:   namespace,
+				ci.Version:        "0",
+			}
+			if k.nodeName != "" {
+				attributes["NodeName"] = k.nodeName
+			}
+			md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+			result = append(result, md)
+		}
+	}
 	return result
 }
 
-func (k *K8sAPIServer) start() error {
-	//TODO: add implementation
+func (k *K8sAPIServer) init() error {
+	var ctx context.Context
+	ctx, k.cancel = context.WithCancel(context.Background())
+
+	k.nodeName = os.Getenv("HOST_NAME")
+	if k.nodeName == "" {
+		return errors.New("missing environment variable HOST_NAME. Please check your deployment YAML config")
```

Comment:
environment HOST_NAME 
```suggestion
		return errors.New("environment variable HOST_NAME is not set in k8s deployment config")
```

Reply:
Will make the change